include/knowhere/index_node.h

```diff
@@ -115,6 +115,72 @@ class IndexNode : public Object {
     Version version_;
 };
 
+// An iterator implementation that accepts a list of distances and ids and returns them in order.
+class PrecomputedDistanceIterator : public IndexNode::iterator {
+ public:
+    PrecomputedDistanceIterator(std::vector<std::pair<float, int64_t>>&& distances_ids, bool larger_is_closer)
+        : comp_(larger_is_closer), results_(std::move(distances_ids)) {
+        sort_size_ = std::max((size_t)50000, results_.size() / 10);
+        sort_next();
+    }
+
+    // Construct an iterator from a list of distances with index being id, filtering out zero distances.
+    PrecomputedDistanceIterator(std::vector<float> distances, bool larger_is_closer) : comp_(larger_is_closer) {
+        for (size_t i = 0; i < distances.size(); i++) {
+            if (distances[i] != 0) {
+                results_.push_back(std::make_pair(distances[i], i));
+            }
+        }
+        sort_size_ = std::max((size_t)50000, results_.size() / 10);
+        sort_next();
+    }
+
+    std::pair<int64_t, float>
+    Next() override {
+        sort_next();
+        auto& result = results_[next_++];
+        return std::make_pair(result.second, result.first);
+    }
+
+    [[nodiscard]] bool
+    HasNext() const override {
+        return next_ < results_.size() && results_[next_].second != -1;
+    }
+
+ private:
+    // sort the next sort_size_ elements
+    inline void
+    sort_next() {
+        if (next_ < sorted_) {
+            return;
+        }
+        size_t current_end = std::min(results_.size(), sorted_ + sort_size_);
+        std::partial_sort(results_.begin() + sorted_, results_.begin() + current_end, results_.end(), comp_);
+        sorted_ = current_end;
+    }
+    struct PairComparator {
+        bool larger_is_closer;
+        PairComparator(bool larger) : larger_is_closer(larger) {
+        }
+
+        bool
+        operator()(const std::pair<float, int64_t>& a, const std::pair<float, int64_t>& b) const {
+            if (a.second == -1) {
+                return false;
+            }
+            if (b.second == -1) {
+                return true;
+            }
+            return larger_is_closer ? a.first > b.first : a.first < b.first;
```

Comment:
what if `a.first == b.first`? Typically, ppl write something like this
````
if (a.first > b.first) return true;
if (a.first < b.first) return false;

// a.first == b.first case
return (a.second > b.second);
```

This is needed to ensure the deterministic behavior.